### PR TITLE
ICU-20342 FormattedDateInterval, Java.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20342
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java [hunk 1]
```diff
@@ -215,54 +221,182 @@
...
+ *
  * @stable ICU 4.0
  */
 
 public class DateIntervalFormat extends UFormat {
 
+    /**
+     * An immutable class containing the result of a date interval formatting operation.
+     *
+     * Not intended for public subclassing.
```
Comment (sffc) [inline icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java]: Add:

```java
 *
 * Instances of this class are immutable and thread-safe.
```

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java [hunk 2]
```diff
@@ -755,76 +971,83 @@ public final synchronized StringBuffer format(Calendar fromCalendar,
...
+        }
         if ( intervalPattern.getSecondPart() != null ) {
             fDateFormat.applyPattern(intervalPattern.getSecondPart());
-            FieldPosition otherPos = new FieldPosition(pos.getField());
-            fDateFormat.format(secondCal, appendTo, otherPos);
-            if (pos.getEndIndex() == 0 && otherPos.getEndIndex() > 0) {
-                pos.setBeginIndex(otherPos.getBeginIndex());
-                pos.setEndIndex(otherPos.getEndIndex());
-            }
+            fDateFormat.format(secondCal, appendTo, pos, attributes);
```
Comment (fabalbon) [inline icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java]: Maybe I'm missing something, but aren't you overriding the value that it's stored in `pos`?
Or will SimpleDateFormat avoid overriding `pos` if `getEndIndex() > 0`?

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/DateIntervalFormatTest.java [hunk 3]
```diff
@@ -1889,4 +1890,60 @@ public void run() {
...
+                {DateIntervalFormat.SpanField.DATE_INTERVAL_SPAN, 21, 34, 0},
+                {DateFormat.Field.MONTH, 21, 25},
+                {DateFormat.Field.DAY_OF_MONTH, 26, 28},
+                {DateFormat.Field.YEAR, 30, 34}};
+            FormattedValueTest.checkFormattedValue(
+                message,
+                result,
+                expectedString,
+                expectedFieldPositions);
+        }
```
Comment (fabalbon) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/DateIntervalFormatTest.java]: Could you add a test case for the time-only fallback pattern? (e.g. "Feb 15, 2 PM – 3 PM")?

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java (hunk 2 above)
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java]: The style I'm using in Java is to unset the reference to the user's `pos` once it is populated.  I believe the first time `pos` could change is in the first call to `fDateFormat.format` on line 1008.  On line 1010, I check if `pos` was updated, and if it was, I set it to a dummy instance.  There were three places I had to do this hack, all with the same comment `// Only accept the first instance of the field`.  Please verify that I caught everywhere I needed to make this change.

In C++, of course, we have FieldPositionHandler, which is a much nicer abstraction to use.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/DateIntervalFormatTest.java (hunk 3 above)
Comment (sffc) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/DateIntervalFormatTest.java]: Done.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/FormattedValueFieldPositionIteratorImpl.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import java.text.Format;
+import java.util.List;
+
+import com.ibm.icu.text.ConstrainedFieldPosition;
+
+/**
+ * Implementation of FormattedValue based on FieldPositionIterator.
+ *
+ * In C++, this implements FormattedValue. In Java, it is a stateless
+ * collection of static functions to avoid having to use nested objects.
```
[the remark this answers is not in the document]
Comment (sffc) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/FormattedValueFieldPositionIteratorImpl.java]: Done.

**Comment by sffc** [comment]
Pushed commits with review feedback.

Let me know if LGTM, then I will squash.

**Comment by sffc** [comment]
The new time test may or may not pass on the build bots due to time zones.  There are some build bots that are in different time zones.

File: icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java (hunk 2 above)
Comment (fabalbon) [reply icu4j/main/classes/core/src/com/ibm/icu/text/DateIntervalFormat.java]: I see, looks good.

**Comment by sffc** [comment]
Looks like the tests are passing.  Made a squash commit.  PTAL.

**Pull request merged**